store/tikv/latch/latch.go

```diff
@@ -0,0 +1,202 @@
+	"sync"
+
+	"github.com/spaolacci/murmur3"
+)
+
+// Latch stores a key's waiting transactions information.
+type Latch struct {
+	// Whether there is any transaction in waitingQueue except head.
+	hasWaiting bool
+	// The startTS of the transaction which is the head of waiting transactions.
+	waitingQueueHead uint64
+	maxCommitTS      uint64
+	sync.Mutex
+}
+
+func (l *Latch) occupied() bool {
+	return l.waitingQueueHead != 0
+}
+
+func (l *Latch) free() {
+	l.waitingQueueHead = 0
+}
+
+// Lock is the locks' information required for a transaction.
+type Lock struct {
+	// The slot IDs of the latches(keys) that a startTS must acquire before being able to processed.
+	requiredSlots []int
+	// The number of latches that the transaction has acquired.
+	acquiredCount int
+	// Whether current transaction is waiting
+	waiting bool
+	// Current transaction's startTS.
+	startTS uint64
+}
+
+// NewLock creates a new lock.
+func NewLock(startTS uint64, requiredSlots []int) Lock {
+	return Lock{
+		requiredSlots: requiredSlots,
+		acquiredCount: 0,
+		waiting:       false,
+		startTS:       startTS,
+	}
+}
+
+// Latches which are used for concurrency control.
+// Each latch is indexed by a slot's ID, hence the term latch and slot are used in interchangeable,
+// but conceptually a latch is a queue, and a slot is an index to the queue
+type Latches struct {
+	slots []Latch
+	// The waiting queue for each slot(slotID => slice of startTS).
+	waitingQueue map[int][]uint64
+	sync.RWMutex
+}
+
+// NewLatches create a Latches with fixed length,
+// the size will be rounded up to the power of 2.
+func NewLatches(size int) *Latches {
+	powerOfTwoSize := 1 << uint32(bits.Len32(uint32(size-1)))
+	slots := make([]Latch, powerOfTwoSize)
+	return &Latches{
+		slots:        slots,
+		waitingQueue: make(map[int][]uint64),
+	}
+}
+
+// GenLock generates Lock for the transaction with startTS and keys.
+func (latches *Latches) GenLock(startTS uint64, keys [][]byte) Lock {
+	slots := make([]int, 0, len(keys))
+	for _, key := range keys {
+		slots = append(slots, latches.hash(key))
+	}
+	sort.Ints(slots)
+	if len(slots) <= 1 {
+		return NewLock(startTS, slots)
+	}
+	dedup := slots[:1]
+	for i := 1; i < len(slots); i++ {
+		if slots[i] != slots[i-1] {
+			dedup = append(dedup, slots[i])
+		}
+	}
+	return NewLock(startTS, dedup)
+}
+
+// hash return hash int for current key.
+func (latches *Latches) hash(key []byte) int {
+	return int(murmur3.Sum32(key)) & (len(latches.slots) - 1)
+}
+
+// Acquire tries to acquire the lock for a transaction.
+// It returns with stale = true when the transaction is stale(
+// when the lock.startTS is smaller than any key's last commitTS).
+func (latches *Latches) Acquire(lock *Lock) (success, stale bool) {
+	for lock.acquiredCount < len(lock.requiredSlots) {
+		slotID := lock.requiredSlots[lock.acquiredCount]
+		success, stale = latches.acquireSlot(slotID, lock.startTS)
+		if success {
+			lock.acquiredCount++
+			lock.waiting = false
+			continue
```

Comment:
Should not `continue` here.
`success` and `stale` may both be true.

Reply:
```
if !stale {
    continue
}
```

Reply:
According to the definition of `acquiredCount`  and `success`, it's ok to continue here.